### ISSUE TITLE
Some additonal experiments

### ISSUE THREAD

**Issue by defgsus** [SciYu/KAE#2]
Hello there,

i've done some additional experiments with your KAE model. The results are published here:

https://defgsus.github.io/nn-experiments/html/logs/review-kan-autoencoder.html

Thanks again, for your research and sharing your code!

**Comment by SciYu** [comment]
Hi,

Thank you for your interest in our work. Your results are insightful and inspire us to improve further. May I kindly know your name? We will update the paper accordingly, particularly regarding the term "superior," and acknowledge your contribution in future revisions.

By the way, could you update our institution name to "_The Chinese University of Hong Kong, Shenzhen_"? Also, if possible, could you share your baseline CNN settings or code? We would like to use them for comparison.

Thanks again for your excellent blog and thoughtful experiments!

Best regards,
Fangchen Yu
On behalf of our research team

**Comment by defgsus** [comment]
Hello @SciYu and all at the team.

Thank you for your kind response :+1: I've updated your institution's name in the post. 
My name is Stefan Berke and it would be really cool to have it on paper for once. I'm not associated with a research institution but the company i work for is [netzkolchose, Germany](https://netzkolchose.de)

Gladly i share the CNN model. It's a stack of convolutions with ReLU activation and one linear layer without activation. I found similar performing models by just using convolution and [pixel-(un)shuffel](https://pytorch.org/docs/stable/generated/torch.nn.PixelShuffle.html) layers, but they result in a different latent size and finding a good model that compresses 28x28 to, e.g., 4x4 would take some time. 

```python
import math
from typing import Tuple

import torch
import torch.nn as nn


class Reshape(nn.Module):

    def __init__(self, shape: Tuple[int, ...]):
        super().__init__()
        self.shape = tuple(shape)

    def forward(self, x):
        return x.reshape(-1, *self.shape)

    def extra_repr(self):
        return f"shape={self.shape}"


class CNNAutoEncoder(nn.Module):

    def __init__(
            self,
            shape: Tuple[int, int, int] = (1, 28, 28),
            latent_dim: int = 16,
            channels: Tuple[int, ...] = (24, 32, 48),
            kernel_size: int = 5,
    ):
        super().__init__()

        self.encoder = nn.Sequential()
        self.decoder = nn.Sequential()

        # --- setup encoder ---

        channels_ = [shape[0], *channels]
        for ch, ch_next in zip(channels_, channels_[1:]):
            self.encoder.append(nn.Conv2d(ch, ch_next, kernel_size))
            self.encoder.append(nn.ReLU())

        # get CNN-stack output size
        with torch.no_grad():
            cnn_out_shape = self.encoder(torch.zeros(1, *shape)).shape[1:]
            cnn_out_dim = math.prod(cnn_out_shape)

        self.encoder.append(nn.Flatten(-3))
        self.encoder.append(nn.Linear(cnn_out_dim, latent_dim))

        # --- setup decoder ---

        self.decoder.append(nn.Linear(latent_dim, cnn_out_dim))
        self.decoder.append(Reshape(cnn_out_shape))

        channels_ = list(reversed(channels_))
        for i, (ch, ch_next) in enumerate(zip(channels_, channels_[1:])):
            is_last_channel = i == len(channels_) - 2
            self.decoder.append(nn.ConvTranspose2d(ch, ch_next, kernel_size))
            if not is_last_channel:
                self.decoder.append(nn.ReLU())

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return self.decoder(self.encoder(x))
```

For the MNIST shape and latent dim of 16 this result in the model:
```
CNNAutoEncoder(
  (encoder): Sequential(
    (0): Conv2d(1, 24, kernel_size=(5, 5), stride=(1, 1))
    (1): ReLU()
    (2): Conv2d(24, 32, kernel_size=(5, 5), stride=(1, 1))
    (3): ReLU()
    (4): Conv2d(32, 48, kernel_size=(5, 5), stride=(1, 1))
    (5): ReLU()
    (6): Flatten(start_dim=-3, end_dim=-1)
    (7): Linear(in_features=12288, out_features=16, bias=True)
  )
  (decoder): Sequential(
    (0): Linear(in_features=16, out_features=12288, bias=True)
    (1): Reshape(shape=(48, 16, 16))
    (2): ConvTranspose2d(48, 32, kernel_size=(5, 5), stride=(1, 1))
    (3): ReLU()
    (4): ConvTranspose2d(32, 24, kernel_size=(5, 5), stride=(1, 1))
    (5): ReLU()
    (6): ConvTranspose2d(24, 1, kernel_size=(5, 5), stride=(1, 1))
  )
)

encoder params: 254,928
decoder params: 267,153
all params: 522,081 
```

When training with AdamW (with pytorch defaults), learnrate 0.0003, batch-size 64 and MSE reconstruction loss, for 10 epochs it yields a test loss of about **0.007**. 

Obviously, the stack of convolutional layers gives the model an advantage in *understanding* the spatial information of the image. It would be interesting to replace the linear layer with your KAELayer. I haven't tried it yet.

Hope you did not take the *superiority* rant personally. It's just something that jumps to my eye more and more. I'm glad i read your paper, nevertheless :smile: 

Here's my email (in base64) if you need it, but we can also further communicate on github 
`cy5iZXJrZUBuZXR6a29sY2hvc2UuZGU=`  

Ciao, and best regards

**Comment by SciYu** [comment]
Hi Stefan,

Thanks a lot for your update and for sharing the code! CNNs do have a natural advantage with image data, so we’re planning to compare their performance and see how the KAELayer works when integrated into CNNs.

I really appreciate your point about overclaimed statements in AI papers—it’s definitely something we aim to stay mindful of and be more rigorous and modest about.

My email is available on my homepage (https://sciyu.github.io/). I'd be happy to continue our conversation over email anytime!

Best,
Fangchen Yu

**Comment by defgsus** [comment]
Hi there, 
in the meantime, i tried a couple of setups with convolutions followed by the KAE layer. There is a small performance gain when putting the KAE layer in the decoder. 
```
CNN-stack -> linear -> latent -> KAE -> CNN-stack -> sigmoid
```
No use to put it in the encoder in my experiments. Interestingly, the sigmoid activation seems to be the best choice after the KAE layer was involved. I could imagine that the polynomial functions work best when they are not constrained to yield a small value range, which would be the case when just using ReLU.

Anyways, i wanted to share a purely CNN/pixel-shuffle model with similar performance to the above mentioned CNN/linear model. The code is in [experiments/ae/scriptae/scriptae2.py @ 3d1e00dc](https://github.com/defgsus/nn-experiments/blob/3d1e00dc208927598dcd82110f6bcc5f464fa7c1/experiments/ae/scriptae/scriptae2.py).

The 28x28 resolution of the MNIST digits do not allow much leeway in the setup to yield a compression ratio of 49 for comparison. Basically, the model does a PixelUnshuffle with factor 7 and then gradually decreases the channel count.

```python
ScriptedAE(
      channels=1,
      kernel_size=3,
      padding=1,
      activation="relu6",
      final_encoder_activation=None,
      final_decoder_activation="relu6",
      batch_norm=False,
      script="ch=32|down7|ch/8|ch/2|ch/2|ch/2|ch/2|ch=1",
)
```

It's a huge model though (6M params), because of the number of channels created in the pixel unshuffeling. It archives l2 test loss of 0.007 after 16 epochs (with similar settings as the CNN/linear model). Using batch-norm increases the train performance but slightly hurts the test performance.

Passing the 1x28x28 shape through the encoder yields the following "stacktrace":

| input           | module            | params                                                               |
|:----------------|:------------------|:---------------------------------------------------------------------|
| (1, 1, 28, 28)  | Sequential        | Sequential()                                                         |
| (1, 1, 28, 28)  | Sequential.0      | ScriptAELayer()                                                      |
| (1, 1, 28, 28)  | Sequential.0.conv | Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))     |
| (1, 32, 28, 28) | Sequential.0.act  | ReLU6()                                                              |
| (1, 32, 28, 28) | Sequential.1      | PixelUnshuffle(downscale_factor=7)                                   |
| (1, 1568, 4, 4) | Sequential.2      | ScriptAELayer()                                                      |
| (1, 1568, 4, 4) | Sequential.2.conv | Conv2d(1568, 196, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1)) |
| (1, 196, 4, 4)  | Sequential.2.act  | ReLU6()                                                              |
| (1, 196, 4, 4)  | Sequential.3      | ScriptAELayer()                                                      |
| (1, 196, 4, 4)  | Sequential.3.conv | Conv2d(196, 98, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))   |
| (1, 98, 4, 4)   | Sequential.3.act  | ReLU6()                                                              |
| (1, 98, 4, 4)   | Sequential.4      | ScriptAELayer()                                                      |
| (1, 98, 4, 4)   | Sequential.4.conv | Conv2d(98, 49, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))    |
| (1, 49, 4, 4)   | Sequential.4.act  | ReLU6()                                                              |
| (1, 49, 4, 4)   | Sequential.5      | ScriptAELayer()                                                      |
| (1, 49, 4, 4)   | Sequential.5.conv | Conv2d(49, 24, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))    |
| (1, 24, 4, 4)   | Sequential.5.act  | ReLU6()                                                              |
| (1, 24, 4, 4)   | Sequential.6      | ScriptAELayer()                                                      |
| (1, 24, 4, 4)   | Sequential.6.conv | Conv2d(24, 12, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))    |
| (1, 12, 4, 4)   | Sequential.6.act  | ReLU6()                                                              |
| (1, 12, 4, 4)   | Sequential.7      | ScriptAELayer()                                                      |
| (1, 12, 4, 4)   | Sequential.7.conv | Conv2d(12, 1, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))     |

resulting in a latent shape of 1x4x4

The nice thing about convolution-only auto-encoders, as you sure know, is the fact that one is not restricted to a fixed resolution. E.g. for this particular model, any resolution that is divisible by 7 can be encoded, always yielding the compression ratio of 49.

Anyways, i wish you all a productive rest of the week and best regards!

**Comment by SciYu** [comment]
Hi Stefan,

Thank you very much for sharing your insights and updates on the CNN+KAN experiments. I sincerely apologize for the delayed response. We have been on our Chinese New Year holiday break, which will conclude in early February.

Once we resume work, we will follow up on the experiments and explore new network architectures based on your valuable input. We look forward to further collaboration and exchanging ideas.

Wishing you a productive week ahead!

**Comment by defgsus** [comment]
Hello again!

I just realized something obvious today :facepalm: For the single layer encoders, i removed the activation function. Usually, in my AE experiments i do not put an activation in front of the latent code. Experiments never show an improvement, using one. 

Below are the results of the single layer models with and without encoder activation. The three new experiments are added to my [KAE fork repo](https://github.com/defgsus/KAE).

| model      | act              |    params | optim/lr/bs      | test loss (10 runs)↓ | classifier accuracy↑ | retriever recall@5↑ | denoiser salt&pepper↓ |
|:-----------|:-----------------|----------:|:-----------------|---------------------:|---------------------:|--------------------:|----------------------:|
| MLP¹       | relu/sigmoid     |    25,888 | Adam/0.0001/256  |       0.0532 ±0.0020 |               0.8859 |              0.4021 |                0.0870 |
| MLP        | relu/relu        |    25,888 | Adamw/0.0003/64  |       0.0294 ±0.0010 |               0.9524 |              0.5261 |                0.0750 |
| MLP        | **none**/relu    |    25,888 | Adamw/0.0003/64  |       0.0230 ±0.0004 |           **0.9638** |          **0.6467** |                0.0702 |
| KAE (p=3)¹ | relu/sigmoid     |   101,152 | Adam/0.0001/256  |       0.0243 ±0.0014 |               0.9523 |              0.5446 |            **0.0672** |
| KAE (p=3)  | relu/sigmoid     |   101,152 | Adam/0.0001/64   |       0.0256 ±0.0020 |               0.9530 |              0.5546 |                0.0676 |
| KAE (p=3)  | **none**/sigmoid |   101,152 | Adam/0.0001/256  |       0.0225 ±0.0003 |               0.9502 |              0.5384 |                0.0673 |
| KAE (p=3)  | **none**/sigmoid |   101,152 | Adamw/0.0003/64  |   **0.0201** ±0.0003 |               0.9518 |              0.5890 |                0.0762 |

(¹: original from paper)

Interestingly, for the no-encoder-activation KAE, the smaller batch-size increases the l2 test performance. 

The MLP without encoder-activation increases performance in classification and retrieval compared to the KAE.

Best regards!
